File: src/org/oopsc/statement/Statement.java
package org.oopsc.statement;

import java.util.Stack;

import org.oopsc.CodeStream;
import org.oopsc.CompileException;
import org.oopsc.SemanticAnalysis;
import org.oopsc.TreeStream;

/**
 * Die abstrakte Basisklasse für alle Anweisungen im Syntaxbaum.
 */
abstract public class Statement {
	public enum Context {
		Default, TryBlock
	}

	/**
	 * Die Methode führt die Kontextanalyse für diese Anweisung durch.
	 *
	 * @param sem
	 *        Die an dieser Stelle gültigen Deklarationen.
	 * @throws CompileException
	 *         Während der Kontextanylyse wurde ein Fehler
	 *         gefunden.
	 */
	public void defPass(SemanticAnalysis sem) throws CompileException {

	}

	public void refPass(SemanticAnalysis sem) throws CompileException {

	}

	/**
	 * Die Methode gibt diese Anweisung in einer Baumstruktur aus.
	 *
	 * @param tree
	 *        Der Strom, in den die Ausgabe erfolgt.
	 */
	abstract public void print(TreeStream tree);

	/**
	 * Die Methode generiert den Assembler-Code für diese Anweisung. Sie geht
	 * davon aus, dass die Kontextanalyse vorher erfolgreich abgeschlossen wurde.
	 *
	 * @param code
	 *        Der Strom, in den die Ausgabe erfolgt.
	 * @param contexts
	 *        Current stack of contexts, may be used to inject instructions for
	 *        unwinding the stack (as needed for RETURN statements in TRY blocks).
	 */
	abstract public void generateCode(CodeStream code, Stack<Context> contexts);
}

File: src/org/oopsc/statement/Statement.scala
package org.oopsc.statement

import org.oopsc.CodeStream
import org.oopsc.SemanticAnalysis
import org.oopsc.TreeStream

abstract class Statement {
  /**
   * Performs the definition pass of the semantic analysis.
   *
   * @param sem Context of the semantic analysis.
   */
  def refPass(sem: SemanticAnalysis) {

  }

  /**
   * Performs the optimisation pass.
   *
   * @return Optimised statement.
   */
  def optimPass() = this

  /**
   * Prints statement in a tree structure.
   *
   * @param tree Output stream.
   */
  def print(tree: TreeStream)

  /**
   * Generates assembly code for the statement. Requires prior completion of the
   * contextual analysis.
   *
   * @param code Output stream.
   * @param tryContexts Current number of TRY blocks. May be used to inject instructions
   *                    for unwinding the stack (as needed for RETURN statements in TRY blocks).
   */
  def generateCode(code: CodeStream, tryContexts: Int)
}